### DISCUSSION TITLE
import error?

### DISCUSSION THREAD

**Discussion by JonTron1987** [remix-run/remix#2730] [Q&A]
Greetings, I am new to Remix. I am currently learning the framework. I am stuck at the routing pages part. I get an error when trying to import from react:

![image](https://user-images.githubusercontent.com/12778501/162772567-1359d001-f190-4323-9143-13e528d0f78a.png)
![image](https://user-images.githubusercontent.com/12778501/162773013-0757904e-3e35-47cd-8074-06627c18786e.png)

I have tried npm add @babel/runtime, deleting node_modules and reinstalling, etc. Nothing appears to work. What can I do to fix the bug?

**Comment by silvenon** [comment]
This looks like an ESLint error, not an error while actually running Remix, correct? What is your ESLint configuration?

**Reply by JonTron1987** [reply]
hm...let me check that out. I am still learning this so that would be in the .eslint folder correct? I seem to have a lot of the same files in different folders...I used the auto-generating npm command to build the template.

**Reply by silvenon** [reply]
You don't have to mark my questions as the answer to your question. I only posted them as an answer because GitHub doesn't yet have the feature to post comments under your question. Let's figure this out. (I'll post comments under your answer below.)

**Reply by JonTron1987** [reply]
ah, I hit that by accident.

**Comment by JonTron1987** [comment]
Okay this is my .eslint.js folder:

![image](https://user-images.githubusercontent.com/12778501/162974272-61da57cf-9d06-4a8d-8275-f2fb4ff7233c.png)

and now it is getting an error in the comments...:
![image](https://user-images.githubusercontent.com/12778501/162974414-29528b1f-afbc-49d3-8d0b-61f50d982ff6.png)

strange because the server still runs perfectly...or so it was. Could this be a bug with visual studio code or Remix?

**Reply by silvenon** [reply]
Ok, so all you did is create a template using `npx create-remix@latest`, no? `@babel/preset-react` should have installed as part of `@remix-run/eslint-config`, because if you check dependencies of that preset:

```
npm info @remix-run/eslint-config dependencies
```

you'll notice `@babel/preset-react` in that list. Are you sure you ran `npm install` in your project?

**Reply by JonTron1987** [reply]
yes, when it asked me to run npm install I told it yes and it ran it.

**Reply by JonTron1987** [reply]
![image](https://user-images.githubusercontent.com/12778501/163174724-c824d3de-ac36-4a0c-8047-de7eeff2ea8c.png)
This is the result of that npm command...it is showing but for some reason VS code is not liking it.

**Reply by JonTron1987** [reply]
ah, now that I look at this closer it may not be there after all...is there a command to force it to install it...I mean I know the --force, but is there another that is sure way to install it?

**Reply by silvenon** [reply]
You shouldn't have to force npm install, that's a red flag that there's something else wrong with your setup that's getting in the way, but it's hard to say what. This message about marking `remix` as external is pretty confusing to me, is the output from the screenshot from when you run `npm run dev`?

**Reply by JonTron1987** [reply]
the output is from using that npm info command

**Reply by JonTron1987** [reply]
the most recent one is from using the npm info command. The error itself is just random; it is happening automatically.

**Reply by silvenon** [reply]
So according to the last output you had to terminate the `npm info` command…? Maybe you don't have npm v8, you can check just in case `npm --version`, perhaps `npm info` works differently in older versions, although that's not really related to the main error it still might help indirectly. Also, you say that VS Code might not be liking your command, I'm pretty sure you'd be getting the same problem inside a separate terminal, but you can try running it in the command line that for example [Git for Windows](https://git-scm.com/download/win) provides.

As for the error you say is happening randomly, it's happening when your VS Code ESLint extension runs ESLint, I assume, so you should be seeing the same error if you ran ESLint outside of VS Code as well, try it just in case:

```
npx eslint app/routes/new.jsx
```

Is using Remix the only time you started running into problems like this? They don't seem related to Remix, I would suggest finding a buddy at work who knows this stuff so that they can poke around your setup, because the problem(s) you're experiencing seem specific to your setup and way easier to debug in person if that's possible 🙂 also, I don't really know what else to ask you 😅

**Reply by JonTron1987** [reply]
Yep, sometimes even with React it does this too...but it disappears eventually too also...but with Remix it stays error on the import part.
Could be the npm too though...been awhile since I updated it.

**Reply by JonTron1987** [reply]
ah...its 8.5.0...

**Reply by JonTron1987** [reply]
yeah I am going to delete the old project and try to run it again. Just to clarify what is the command to create just a basic default template?

**Reply by silvenon** [reply]
Are you asking about this one?

```
npx create-remix@latest
```

It's going to ask you a question whether you want to use one of the stacks (Indie, Blues etc.), and you select that you don't want to, just the basics, then it's going to ask where do you plan to deploy it, if you don't know yet I recommend selecting the first choice (Remix app server I think it's called).

**Reply by silvenon** [reply]
I hope it goes well this time! 🤞 🤞 🤞

**Reply by JonTron1987** [reply]
Sorry for the late reply. Yes I think it was were ever I got that command from; it was updated recently so I had to change the command.

**Reply by silvenon** [reply]
I didn't quite understand, is it working now…?

**Reply by camjackson** [reply]
@silvenon I'm getting a very similar error to JonTron's one. I'm still debugging it but here's what I know so far:
- everything was fine on remix 1.4.3
- after upgrading all remix packages (including the eslint-config package) to 1.6.0, I start getting this `Cannot find module '@babel/preset-react'` error from eslint
- `yarn why @babel/preset-react` gives the following output:
```
=> Found "@babel/preset-react@7.17.12"
info Reasons this module exists
   - "@remix-run#eslint-config" depends on it
   - Hoisted from "@remix-run#eslint-config#@babel#preset-react"
```
- I can confirm that there is no such `node_modules/@babel/preset-react` directory, if I fire up a node repl in my project directory and try to `require('@babel/preset-react')`, I get a `MODULE_NOT_FOUND` error

So it seems that the dependency is genuinely not there in my node_modules, despite yarn knowing that it's a transitive dependency of the remix eslint config. Very mysterious. I might need to bisect the eslint config package versions and see if I can figure out exactly which version in between 1.4.3 and 1.6.0 introduces the problem.

**Reply by camjackson** [reply]
Small update: 1.5.1 is good, 1.6.0 is bad. These are all the packages I'm switching the versions of:
- @remix-run/dev
- @remix-run/eslint-config
- @remix-run/architect
- @remix-run/node
- @remix-run/react
- @remix-run/server-runtime

I'll try to see if I can narrow it down to a single package.

**Reply by camjackson** [reply]
Update: It's `@remix-run/dev`. With all the other packages on 1.5.1, I can see a `node_modules/@babel/preset-react` directory and eslint functions just fine. As soon as I upgrade to `@remix-run/dev@1.5.1`, `node_modules/@babel/preset-react` disappears and eslint no longer works. Will keep digging.

**Reply by camjackson** [reply]
For the life of me I cannot figure out why upgrading that one package version causes the babel preset to go missing from node_modules despite still being listed in the yarn.lock. It's bizarre. I do know that when I explicitly install it by doing `yarn add -D @babel/preset-react`, the problem goes away. So that's at least a work-around for now.

**Reply by silvenon** [reply]
Hm, AFAIK `@babel/preset-react` doesn't have to be hoisted to work, i.e. it doesn't have to be in `node_modules/@babel/preset-react`, hoisting happens only if same dependencies of different packages happen to match in versions. `@babel/preset-react` can be in `node_modules/@remix-run/eslint-config/node_modules/@babel/preset-react` too, because `@remix-run/eslint-config` is the package that's using it.

I'm not sure what could be the issue, try running `npx eslint <path to your file>` to check whether the problem is only in your editor's ESLint extension or really in your ESLint configuration. You need to run that in the directory where `node_modules` is. Also, do you by any chance have a monorepo? And what is your version of Yarn?

**Reply by camjackson** [reply]
Ah good point, I'm forgetting that it doesn't need to be hoisted and that it can just be within the `node_modules` of the package that needs it.

It's definitely not an editor issue, I've been running eslint from my terminal using `yarn lint`, which does this command:

```
eslint --cache --cache-location ./node_modules/.cache/eslint .
```

I'm running node v18.0.0, and yarn 1.22.17. I'm not using the newer plug'n'play or zero-install features, and I'm not using yarn workspaces, lerna or any other monorepo tool.

Here's a theory... What if there's some library somewhere in the dependency tree which needs `@babel/preset-react`, but has not listed it as a dependency? It could be that:

Before the upgrade:
- Two different packages depended on the same version of `@babel/preset-react`
- Therefore, `@babel/preset-react` got hoisted into the top-most `node_modules`
- So the offending library was still able to load its missing dependency

After the upgrade:
- Every package that depends on `@babel/preset-react` needs a different version
- Therefore, `@babel/preset-react` is not hoisted, each package gets its own local copy in its nested `node_modules`
- So the offending library now fails to load its missing dependency

That could be it? It would need a bit of untangling to confirm if that's the problem and then to find the culprit if that is the case.

**Reply by silvenon** [reply]
Your setup sounds standard and straightforward. And regarding the possibility about a library which doesn't list it as a dependency, I don't see how that would case this error. 😔 

Does this happen in a new Remix project, or is there something specific to your setup?

**Reply by camjackson** [reply]
> I don't see how that would case this error.

I was thinking there could be a library that that is importing the babel preset without listing it as a dependency. So it's basically an un-declared peer dependency. It will work if the babel preset happens to be in your top-level node_modules anyway, either because a) you install the babel preset as a direct dependency, which is what I've done, or b) you have some other dependencies that depend on the babel preset in such a way that it gets hoisted.

If neither of those things is true then the babel preset won't be in the node_modules and so the offending library will get an import error.

I created a fresh grunge app a couple of months ago, with an older remix version (1.4 I think) and it wasn't a problem. It was only upgrading to remix 1.6 that it started happening. I've just created a fresh grunge app, which is now on 1.6.1 by default, and it works fine. And the babel preset has been hoisted again. 

Given that it seems to work fine out of the box now, it's probably not really worth anyone spending any more time on this unless a bunch of people start reporting it. It must just be some pathological combination of dependencies that causes it, and there's a simple enough work-around.

**Reply by silvenon** [reply]
> Given that it seems to work fine out of the box now, it's probably not really worth anyone spending any more time on this unless a bunch of people start reporting it. It must just be some pathological combination of dependencies that causes it, and there's a simple enough work-around.

You're right, I just get really curious about these things. 🤓 It's great that you found a workaround!

**Reply by camjackson** [reply]
Yeah I would love to know what the root cause is, but my patience for investigating it further has run out 😅

**Comment by JonTron1987** [comment]
Ah, yeah that is when it started for me too...I kind of figured that was the case, but because I am a beginner and learning Remix I was not certain for sure. Oh, I have a suggestion to the devs who made this; it would be cool if you guys can do a live twitch or Youtube howto for beginners Remix. The remix is a little different from React. Some people get more from actually coding alongside someone than by reading documentation. I am one of them. There is very little streams on this framework right now as it is new correct?    On Wednesday, June 22, 2022, 09:49:26 AM EDT, Cam Jackson ***@***.***> wrote:  
 
 


For the life of me I cannot figure out why upgrading that one package version causes the babel preset to go missing from node_modules despite still being listed in the yarn.lock. It's bizarre. I do know that when I explicitly install it by doing yarn add -D @babel/preset-react, the problem goes away. So that's at least a work-around for now.
